File: backend/app/api/v1/endpoints/result.py
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.core.database import get_db
from app.domain.models import IntakeSession, Recommendation
from app.domain.schemas.patient import (
    InputSummary,
    PackageRecommendation,
    QuestionnaireResponse,
    RedFlagResult,
)

router = APIRouter()


@router.get("/result/{session_key}", response_model=QuestionnaireResponse)
async def get_result(
    session_key: str,
    db: AsyncSession = Depends(get_db),
) -> QuestionnaireResponse:
    try:
        uuid.UUID(session_key)
    except ValueError:
        raise HTTPException(status_code=404, detail="문진 결과를 찾을 수 없습니다")

    stmt = (
        select(IntakeSession)
        .where(IntakeSession.session_key == session_key)
        .options(
            selectinload(IntakeSession.recommendations).selectinload(Recommendation.package)
        )
    )
    result = await db.execute(stmt)
    session = result.scalar_one_or_none()

    if session is None:
        raise HTTPException(status_code=404, detail="문진 결과를 찾을 수 없습니다")

    red_flag_details = session.red_flag_details or {}
    red_flag = RedFlagResult(
        level=session.red_flag_level,
        matched_rules=red_flag_details.get("matched_rules", []),
        message=red_flag_details.get("message", ""),
    )

    input_summary_data = session.input_summary or {}
    input_summary = InputSummary(
        age=input_summary_data.get("age", session.age),
        gender=input_summary_data.get("gender", session.gender),
        symptoms=input_summary_data.get("symptoms", []),
        duration=input_summary_data.get("duration", session.duration),
        existing_conditions=input_summary_data.get("existing_conditions", []),
    )

    sorted_recs = sorted(session.recommendations, key=lambda r: r.rank)
    recommendations = [
        PackageRecommendation(
            package_id=rec.package_id,
            package_name=rec.package.name if rec.package else "",
            match_score=float(rec.match_score),
            reason=rec.reason or "",
            matched_tags=rec.matched_tags or [],
        )
        for rec in sorted_recs
    ]

    return QuestionnaireResponse(
        session_key=str(session.session_key),
        summary=session.llm_summary,
        input_summary=input_summary,
        red_flag=red_flag,
        recommendations=recommendations,
    )

File: frontend/tests/fixtures/result.ts
import type { QuestionnaireResponse } from "@/types";

export const mockResultNone: QuestionnaireResponse = {
  session_key: "test-session-key",
  summary: "45세 남성 환자가 두통과 피로감을 호소합니다.",
  input_summary: {
    age: 45,
    gender: "M",
    symptoms: ["HEADACHE", "FATIGUE"],
    duration: "1~2주",
    existing_conditions: ["고혈압"],
  },
  red_flag: {
    level: "NONE",
    matched_rules: [],
    message: "",
  },
  recommendations: [
    {
      package_id: 1,
      package_name: "기본 종합검진",
      match_score: 0.85,
      reason: "두통과 피로감에 대한 기본 검사",
      matched_tags: ["HEADACHE", "FATIGUE"],
    },
    {
      package_id: 2,
      package_name: "뇌신경 검진",
      match_score: 0.72,
      reason: "두통 증상에 특화된 검진",
      matched_tags: ["HEADACHE"],
    },
  ],
};

export const mockResultCaution: QuestionnaireResponse = {
  ...mockResultNone,
  red_flag: {
    level: "CAUTION",
    matched_rules: ["만성 기침 + 체중 감소"],
    message: "주의가 필요한 증상 조합이 감지되었습니다.",
  },
};

export const mockResultUrgent: QuestionnaireResponse = {
  ...mockResultNone,
  red_flag: {
    level: "URGENT",
    matched_rules: ["편측 마비 + 언어장애"],
    message: "빠른 시일 내에 전문의 진료를 받으시기 바랍니다.",
  },
};

export const mockResultEmergency: QuestionnaireResponse = {
  ...mockResultNone,
  red_flag: {
    level: "EMERGENCY",
    matched_rules: ["흉통 + 팔 방사통"],
    message: "즉시 가까운 응급실을 방문해주세요.",
  },
};
